File: barretenberg/test/wire_compat/ts/echo_server.ts
/**
 * Echo IPC server (TypeScript) — uses GENERATED types and dispatch.
 * Usage: npx tsx echo_server.ts --socket /tmp/echo.sock
 */

import * as net from 'node:net';
import * as fs from 'node:fs';
import { Decoder, Encoder } from 'msgpackr';
import type { Handler } from './generated/server.js';
import { dispatch } from './generated/server.js';
import type { EchoBytes, EchoBytesResponse, EchoFields, EchoFieldsResponse, EchoNested, EchoNestedResponse } from './generated/api_types.js';

const encoder = new Encoder({ useRecords: false, int64AsNumber: true });
const decoder = new Decoder({ useRecords: false, int64AsNumber: true });

const args = process.argv.slice(2);
const socketIdx = args.indexOf('--socket');
const socketPath = socketIdx >= 0 ? args[socketIdx + 1] : undefined;
if (!socketPath) {
  console.error('Usage: echo_server.ts --socket <path>');
  process.exit(1);
}

try { fs.unlinkSync(socketPath); } catch {}

// Implement the GENERATED Handler interface
const handler: Handler = {
  async echoBytes(cmd: EchoBytes): Promise<EchoBytesResponse> {
    return { data: cmd.data };
  },
  async echoFields(cmd: EchoFields): Promise<EchoFieldsResponse> {
    return { a: cmd.a, b: cmd.b, name: cmd.name };
  },
  async echoNested(cmd: EchoNested): Promise<EchoNestedResponse> {
    return { inner: cmd.inner };
  },
};

const server = net.createServer((conn) => {
  let buffer = Buffer.alloc(0);

  conn.on('data', (data: Buffer) => {
    buffer = Buffer.concat([buffer, data]);

    while (buffer.length >= 4) {
      const len = buffer.readUInt32LE(0);
      if (buffer.length < 4 + len) break;

      const payload = buffer.subarray(4, 4 + len);
      buffer = buffer.subarray(4 + len);

      const request = decoder.unpack(payload) as any[];
      const [commandName, fields] = request[0] as [string, any];

      if (commandName === 'EchoShutdown') {
        sendResponse(conn, ['EchoShutdownResponse', {}]);
        setTimeout(() => {
          server.close();
          try { fs.unlinkSync(socketPath); } catch {}
          console.error('echo_server(ts): shutdown');
          process.exit(0);
        }, 50);
        return;
      }

      // Use GENERATED dispatch function
      dispatch(handler, commandName, fields).then(([respName, respFields]) => {
        sendResponse(conn, [respName, respFields]);
      }).catch(err => {
        sendResponse(conn, ['EchoErrorResponse', { message: err.message }]);
      });
    }
  });
});

function sendResponse(conn: net.Socket, response: [string, any]) {
  const packed = encoder.pack(response);
  const lenBuf = Buffer.alloc(4);
  lenBuf.writeUInt32LE(packed.length, 0);
  conn.write(lenBuf);
  conn.write(packed);
}

server.listen(socketPath, () => {
  console.error(`echo_server(ts): listening on ${socketPath}`);
});

File: barretenberg/test/wire_compat/rust/src/echo_server.rs
//! Echo IPC server — echoes commands back as responses.
//! Usage: echo_server --socket /tmp/echo.sock

use echo_wire_compat::generated_types::*;
use echo_wire_compat::server::Handler;
use echo_wire_compat::error::{EchoError, Result};
use std::io::{Read, Write};
use std::os::unix::net::UnixListener;

struct EchoHandler;

impl Handler for EchoHandler {
    fn bytes(&mut self, cmd: EchoBytes) -> Result<EchoBytesResponse> {
        Ok(EchoBytesResponse { data: cmd.data })
    }
    fn fields(&mut self, cmd: EchoFields) -> Result<EchoFieldsResponse> {
        Ok(EchoFieldsResponse { a: cmd.a, b: cmd.b, name: cmd.name })
    }
    fn nested(&mut self, cmd: EchoNested) -> Result<EchoNestedResponse> {
        Ok(EchoNestedResponse { inner: cmd.inner })
    }
}

fn main() -> Result<()> {
    let args: Vec<String> = std::env::args().collect();
    let socket_path = args.iter()
        .position(|a| a == "--socket")
        .and_then(|i| args.get(i + 1))
        .expect("Usage: echo_server --socket <path>");

    // Remove stale socket
    let _ = std::fs::remove_file(socket_path);

    let listener = UnixListener::bind(socket_path)?;
    eprintln!("echo_server(rust): listening on {}", socket_path);

    let (mut stream, _) = listener.accept()?;
    let mut handler = EchoHandler;

    loop {
        // Read 4-byte LE length
        let mut len_buf = [0u8; 4];
        if stream.read_exact(&mut len_buf).is_err() {
            break;
        }
        let len = u32::from_le_bytes(len_buf) as usize;

        // Read payload
        let mut payload = vec![0u8; len];
        stream.read_exact(&mut payload)?;

        // Deserialize: [[CommandName, {fields}]]
        let request: Vec<Command> = rmp_serde::from_slice(&payload)
            .map_err(|e| EchoError::Deserialization(e.to_string()))?;

        let command = request.into_iter().next()
            .ok_or_else(|| EchoError::Deserialization("empty request".into()))?;

        // Check for shutdown
        let is_shutdown = matches!(&command, Command::EchoShutdown(_));

        // Dispatch
        let response = match echo_wire_compat::server::dispatch(&mut handler, command) {
            Ok(resp) => resp,
            Err(e) => Response::EchoErrorResponse(EchoErrorResponse {
                message: e.to_string(),
            }),
        };

        // Serialize response
        let response_bytes = rmp_serde::to_vec_named(&response)
            .map_err(|e| EchoError::Serialization(e.to_string()))?;

        // Send length-prefixed response
        let resp_len = (response_bytes.len() as u32).to_le_bytes();
        stream.write_all(&resp_len)?;
        stream.write_all(&response_bytes)?;
        stream.flush()?;

        if is_shutdown {
            break;
        }
    }

    let _ = std::fs::remove_file(socket_path);
    eprintln!("echo_server(rust): shutdown");
    Ok(())
}
